src/storage/redis_db.cc

```diff
@@ -690,4 +693,61 @@ Status WriteBatchLogData::Decode(const rocksdb::Slice &blob) {
 
   return Status::OK();
 }
+
+rocksdb::Status Database::Rename(const std::string &key, const std::string &new_key, bool nx, bool *ret) {
+  *ret = true;
+  RedisType type = kRedisNone;
+  std::string ns_key = AppendNamespacePrefix(key);
+  std::string new_ns_key = AppendNamespacePrefix(new_key);
+  std::vector<std::string> lock_keys = {ns_key, new_ns_key};
+  MultiLockGuard guard(storage_->GetLockManager(), lock_keys);
+  auto s = Type(key, &type);
+  if (!s.ok()) return s;
+  if (type == kRedisNone) return rocksdb::Status::InvalidArgument("ERR no such key");
+
+  if (nx) {
+    int exist = 0;
+    if (s = Exists({new_key}, &exist), !s.ok()) return s;
+    if (exist > 0) {
+      *ret = false;
+      return rocksdb::Status::OK();
+    }
+  }
+
+  if (key == new_key) return rocksdb::Status::OK();
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(type);
+  batch->PutLogData(log_data.Encode());
+
+  engine::DBIterator iter(storage_, rocksdb::ReadOptions());
+  iter.Seek(ns_key);
+
+  // copy metadata
+  batch->Delete(metadata_cf_handle_, ns_key);
+  batch->Put(metadata_cf_handle_, new_ns_key, iter.Value());
+
+  auto subkey_iter = iter.GetSubKeyIterator();
+
+  if (subkey_iter != nullptr) {
+    for (subkey_iter->Seek(); subkey_iter->Valid(); subkey_iter->Next()) {
+      InternalKey from_ikey(subkey_iter->Key(), storage_->IsSlotIdEncoded());
+      std::string to_ikey =
+          InternalKey(new_ns_key, from_ikey.GetSubKey(), from_ikey.GetVersion(), storage_->IsSlotIdEncoded()).Encode();
+      // copy sub key
+      batch->Put(subkey_iter->ColumnFamilyHandle(), to_ikey, subkey_iter->Value());
+
+      if (type == kRedisZSet) {
+        std::string score_bytes = subkey_iter->Value().ToString();
+        score_bytes.append(from_ikey.GetSubKey().ToString());
+        // copy score key
+        std::string score_key =
+            InternalKey(new_ns_key, score_bytes, from_ikey.GetVersion(), storage_->IsSlotIdEncoded()).Encode();
+        batch->Put(storage_->GetCFHandle(engine::kZSetScoreColumnFamilyName), score_key, Slice());
```

Comment:
Is there a cleaner way to avoid repetitive calls to the GetCFHandle function?
Or, this way is ok, as CPU branch prediction will help us deal with it.

Reply:
Sure. The one way is to just make it before the loop.